File: solutions/haskell/grains/1/src/Grains.hs
module Grains (square, total) where

pow2 :: Integer -> Integer
pow2 n = 2 ^ (n - 1)

square :: Integer -> Maybe Integer
square n
 | n <= 0 || n > 64 = Nothing
 | otherwise        = Just (pow2 n)

total :: Integer
total = sum $ fmap pow2 [1 .. 64]

File: solutions/erlang/grains/1/src/grains.erl
-module(grains).
-export([square/1, total/0]).

pow2(N) -> trunc(math:pow(2,N)).

square(N) when N < 1; N > 64 ->
  {error, "square must be between 1 and 64"};
square(N) -> pow2(N-1).

total() -> pow2(64)-1.
